### ISSUE TITLE
new Serializer API does not load relationships if included record does not have reflexive entry.

### ISSUE THREAD

**Issue by Kilowhisky** [emberjs/data#3614]
In previous versions of ember data the reflexive relationship wasn't required. But with the new serializer API it appears if you return included relationships from `normalizeSingleResponse` without the reflexive like so.

```
{
    "data":{
         "id":"eup8h",
         "type":"item",
          "attributes":{
                "name":"three thousand"
          },
      "relationships":{
             "list":{
                 "data":{
                     "id":"l2",
                     "type":"list"
                  }
               }
       }
   },
   "included":[
         {
             "id":"l2",
              "type":"list",
               "attributes":{
                  "name":"two",
                   "b":false
                },
               "relationships":{
                      "items":{
                           "data":[
                             ]
                       }
               }
          }
     ]
}
```

`item.get('list')` will return null even though the record was loaded with the correct relationship. If you include the inverse relationship under the `included[]` `item.get('list')` will return the included object.  

aka. `included:[{relationships:items{data[{ id: eup8h, type: 'item' }]}}]`

For instance this test will fail at the debugger statement

```
test('saves belongsTo', function() {
  var item,
      listId = 'l2';

  stop();
  store.findRecord('list', listId).then(function(list) {
    item = store.createRecord('item', { name: 'three thousand' });
    item.set('list', list);

    return item.save();
  }).then(function(item) {
    store.unloadAll('item');
    return store.findRecord('item', item.get('id'));
  }).then(function(item) {
    // TODO: THIS IS NOT RECOVERING THE LIST.
    debugger;
    var list = item.get('list');
    ok(item.get('list'), 'list is present');
    equal(list.id, listId, 'list is retrieved correctly');
    start();
  });
});
```

with these model definitions.

```
    App.List = DS.Model.extend({
      name: DS.attr('string'),
      b: DS.attr('boolean'),
      items: DS.hasMany('item'),
    });

    App.Item = DS.Model.extend({
      name: DS.attr('string'),
      list: DS.belongsTo('list')
    });
```

And this fixture data
```
  'list': {
    records: {
      'l1': { id: 'l1', name: 'one', b: true, items: ['i1', 'i2'] },
      'l2': { id: 'l2', name: 'two', b: false, items: [] },
      'l3': { id: 'l3', name: 'three', b: false, items: [] }
    }
  },
```

**Comment by wecc** [comment]
Both this:

```
{
    "data":{
         "id":"eup8h",
         "type":"item",
          "attributes":{
                "name":"three thousand"
          },
      "relationships":{
             "list":{
                 "data":{
                     "id":"l2",
                     "type":"list"
                  }
               }
       }
   },
   "included":[
         {
             "id":"l2",
              "type":"list",
               "attributes":{
                  "name":"two",
                   "b":false
                },
               "relationships":{
                      "items":{
                           "data":[
                             ]
                       }
               }
          }
     ]
}
```

and this:

```
  'list': {
    records: {
      'l1': { id: 'l1', name: 'one', b: true, items: ['i1', 'i2'] },
      'l2': { id: 'l2', name: 'two', b: false, items: [] },
      'l3': { id: 'l3', name: 'three', b: false, items: [] }
    }
  }
```

...gives ED conflicting information. First you're telling ED there's something in the relationship and then that there's nothing. I suspect this working/not working depends on what order records gets added to the store.

You are not required to provide data for relationships in included/sideloaded resources, but then you have to omit it. See http://emberjs.jsbin.com/surulu/edit?html,js,output for an example using JSON API.

**Comment by Kilowhisky** [comment]
I can't quite omit it if its what the server believes at the time as it wouldn't reflect the correct state.  
Changing the behavior based on the order seems like smell to me as well.  

For instance if the i1 item record was returned with belongsTo of [l1,l2,l3] and the list item l1 was returned with a hasMany of [i2,i3,i4] is l1 referenced properly from i1 and vice versa?   Since ember data doesn't save referenced records at the same time then its perfectly possible scenario to occur as shown in the test above where 1 record was saved and the other not.  filling out the reflexive property shouldn't be a requirement on the server since it wasn't before.

**Comment by wecc** [comment]
> I can't quite omit it if its what the server believes at the time as it wouldn't reflect the correct state.

I'm not following. Are we talking about different requests at different times? Your first JSON API example looked like a compound document to me, sent form the server to the client at a single point in time?

**Comment by Kilowhisky** [comment]
Its a compound document that reflects what the server currently thinks the state is. Per your description it sounds like i now need to modify the response so ember data can make the correct connections, correct?   If that is the case after i make the changes now ember data's state no longer reflects the servers since i have added a connection from list to item that wasn't present before. (or omitted connections at all so that ember data doesn't get confused).

**Comment by wecc** [comment]
> Its a compound document that reflects what the server currently thinks the state is.

But your examples provides two different truths within the same response?
Would you be able to provide a simple JSBin demonstrating the issue? That will help making this clearer.

Feel free to ping me in the Ember Community Slack if you're in there.

**Comment by Kilowhisky** [comment]
I'll try and create a JSBin but since i started ember with ember-cli i'm not familiar with the single page syntax.  

Until then i have the repo i'm trying to update located here: https://github.com/Kilowhisky/ember-localstorage-adapter/tree/ember-1-13

You can run the tests and see the test that fails. 


As you say they are conflicting truths. What should the result of the test i described above be? Because with the sideloading mechanism that was standard before the new api it passed and now it fails. 

Here is the old extractSingle

```
extractSingle: function(store, type, payload) {
      if (payload && payload._embedded) {
        for (var relation in payload._embedded) {
          var relType = type.typeForRelationship(relation,store);
          var typeName = relType.modelName,
              embeddedPayload = payload._embedded[relation];

          if (embeddedPayload) {
            if (Ember.isArray(embeddedPayload)) {
              store.pushMany(typeName, embeddedPayload);
            } else {
              store.push(typeName, embeddedPayload);
            }
          }
        }

        delete payload._embedded;
      }

      return this.normalize(type, payload);
    },
```

Here is the new one.
```
normalizeSingleResponse: function(store, type, payload) {
      var included = [];

      if (payload && payload._embedded) {
        for (var relation in payload._embedded) {
          var relType = type.typeForRelationship(relation,store);
          var typeName = relType.modelName,
              embeddedPayload = payload._embedded[relation];

          if (embeddedPayload) {
            if (Ember.isArray(embeddedPayload)) {
              embeddedPayload.forEach(function(record) {
                included.pushObject(this.normalize(relType,record).data);
              }.bind(this));
            } else {
              included.pushObject(this.normalize(relType, embeddedPayload).data);
            }
          }
        }

        delete payload._embedded;
      }

      var normalPayload = this.normalize(type, payload);
      if(included.length > 0){
        normalPayload['included'] = included;
      }
      return normalPayload;
    },
```

**Comment by sebyx07** [comment]
yep definitely need this for pagination

**Comment by wecc** [comment]
@Kilowhisky 
I'm still failing to see what the real issue here is. I would gladly help but I need some more information.

The JSBin is unfortunately way to hard to follow along and it fails with "Ember is not defined" :( If you would be able to create a minimal use case replicating the issue that would be really awesome. I've created a template you can use to get started: http://emberjs.jsbin.com/qefeda/1/edit?html,js,output

Just to make sure, is this issue about the built-in serializers in Ember Data or the LocalStore Adapter?

**Comment by fivetanley** [comment]
Closing due to lack of response. Feel free to reopen or open with more information.
